staging/src/k8s.io/apiserver/pkg/registry/generic/registry/corrupt_obj_deleter.go

```diff
@@ -0,0 +1,139 @@
+	}
+
+	key, err := d.store.KeyFunc(ctx, name)
+	if err != nil {
+		return nil, false, err
+	}
+	obj := d.store.NewFunc()
+	qualifiedResource := d.store.qualifiedResourceFromContext(ctx)
+	// use the storage implementation directly, bypass the dryRun layer
+	storageBackend := d.store.Storage.Storage
+	// we leave ResourceVersion as empty in the GetOptions so the
+	// object is retrieved from the underlying storage directly
+	err = storageBackend.Get(ctx, key, storage.GetOptions{}, obj)
+	if err == nil || !storage.IsCorruptObject(err) {
+		// TODO: The Invalid error should have a field for Resource.
+		// After that field is added, we should fill the Resource and
+		// leave the Kind field empty. See the discussion in #18526.
+		qualifiedKind := schema.GroupKind{Group: qualifiedResource.Group, Kind: qualifiedResource.Resource}
+		fieldErrList := field.ErrorList{
+			field.Invalid(field.NewPath("ignoreStoreReadErrorWithClusterBreakingPotential"), true, "is exclusively used to delete corrupt object(s), try again by removing this option"),
+		}
+		return nil, false, apierrors.NewInvalid(qualifiedKind, name, fieldErrList)
+	}
+
+	// try normal deletion anyway, it is expected to fail
+	obj, deleted, normalDeleteErr := d.store.Delete(ctx, name, deleteValidation, opts)
+	if normalDeleteErr == nil {
+		return obj, deleted, nil
+	}
+	if !strings.Contains(normalDeleteErr.Error(), "corrupt object") {
+		// TODO: unfortunately we can't do storage.IsCorruptObject(err),
+		// conversion to API error drops the inner error chain
+		return obj, deleted, normalDeleteErr
+	}
+
+	var (
+		preconditions *storage.Preconditions
+		internalErr   storage.InternalError
+	)
+	// if we have the resource version of the object then we pin it to the
+	// preconditions, otherwise we drop preconditions entirely.
+	if errors.As(err, &internalErr) && internalErr.ResourceVersion != 0 {
+		preconditions = &storage.Preconditions{
+			ResourceVersion: ptr.To[string](strconv.FormatInt(internalErr.ResourceVersion, 10)),
+		}
+	}
+
+	// TODO: at this instant, some actor may have a) managed to recreate this
+	// object by doing a delete+create, or b) the underlying error has resolved
+	// since the last time we checked, and the object is readable now.
+	// if precondition is set, then we can prevent a
+	klog.FromContext(ctx).V(1).Info("Going to perform unsafe object deletion", "object", klog.KRef(genericapirequest.NamespaceValue(ctx), name))
+	audit.AddAuditAnnotation(ctx, "apiserver.k8s.io/delete-ignore-read-error", "")
+
+	out := d.store.NewFunc()
+	storageOpts := storage.DeleteOptions{IgnoreStoreReadError: true}
+	// keep the admission
+	if err := storageBackend.Delete(ctx, key, out, preconditions, storage.ValidateObjectFunc(deleteValidation), nil, storageOpts); err != nil {
+		if storage.IsNotFound(err) {
+			// the DELETE succeeded, but we don't have the object since it's
+			// not retrievable from the storage, so we send a nil object
+			return nil, false, nil
```

Comment:
I haven't finished reviewing so it may already exist, but we need to see what currently well behaved informers do with this.  In particular, what's in my head is the resourceVersion on a watch.  I remember there was (and still is I think) an issue where events can be deleted in the same resourceVersion.  I'd like to find a way prove that the deletion has a unique resourceVersion.

Reply:
as far as response to the delete request, for a normal deletion flow, we return the old object with a more recent revision:
https://github.com/kubernetes/kubernetes/blob/c19ffb7e72d8a2064f767a61acd63694dc76a6da/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go#L368

so for a watcher watching this object, it will see these events:
- a) `ADDED`, RV: x 
- b) `DELETED` RV: y
for both a and b, it's the same object in `origState.data`, but y > x because of `txnResp.Revision`

However for unsafe deletion, since we can not transform or decode, we return a `nil` object to the response of the delete request.

so for a watcher watching this object, it will see these events:
- a) `ADDED`, RV: x 
- b) instead of a `DELETED` event, it will see an `ERROR` event because of transform or decode failure
https://github.com/kubernetes/kubernetes/blob/c19ffb7e72d8a2064f767a61acd63694dc76a6da/staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go#L681-L697

https://github.com/kubernetes/kubernetes/blob/c19ffb7e72d8a2064f767a61acd63694dc76a6da/staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go#L564-L569

https://github.com/kubernetes/kubernetes/blob/c19ffb7e72d8a2064f767a61acd63694dc76a6da/staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go#L636-L646

`prepareObjs` returns an error like this:
```
failed to prepare current and previous objects: data from the storage is not transformable: no matching prefix found
```
and it causes the watcher to terminate with an `watch.Error` event, and the reflector sees this as an internal error and retries the watch
https://github.com/kubernetes/kubernetes/blob/c19ffb7e72d8a2064f767a61acd63694dc76a6da/staging/src/k8s.io/client-go/tools/cache/reflector.go#L488-L490

> I remember there was (and still is I think) an issue where events can be deleted in the same resourceVersion

not sure what events are you referring to here, events as in event resource?

Reply:
> b) instead of a DELETED event, it will see an ERROR event because of transform or decode failure

we can try to construct an empty object, fill it with the namespace, name, and the revision after delete, this will result in construction of a watch.Deleted event with a partial object. This is hacky, and other implications, what if a workload is generating keys for the store based on other attributes of the object like from the labels or annotations

Reply:
@deads2k I have a solution at https://github.com/kubernetes/kubernetes/pull/127513/commits/46c3fdc4eeda8d3ffb1839fff196c5fece75066d let me know if this is palatable, I can bring it up for discussion.

Reply:
What happens if `handleUnsafeDeleteEventError` is not called yet because the client library is old in some component? Will an informer eventually do the right thing? Is there a way to prove that?

Reply:
@sttts we have opted to implement option A from https://github.com/kubernetes/enhancements/pull/4949

https://github.com/kubernetes/kubernetes/pull/127513/commits/0f4dd0bd5c93865b0d46cbc4710e31a313f5cc89 does not introduce any change in the client, so existing/old clients would work just fine.

does this answer the question you have?

Reply:
yes, with option A it does. The referenced commit above is not part of the PR.